Do not override schema defaults if they exist

jdavcs opened the pull request:
Instead of `_defaults_to('some val')`, use schema default.

The following default values have been changed to use schema
defaults (I assume that was the intent):

key: old default / schema default

genomespace_ui_url: none / https://gsui.genomespace.org/jsui/
communication_server_port: none / 7070
communication_server_host: none / http://localhost
persistent_communication_rooms: none / []
require_login: none / false
inactivity_box_content: none / schema text
registration_warning_message: none  / schema text
welcome_url: none / /static/welcome.html

I did not modify default values if the schema default was the opposite (true vs
false) or if it was overwritten in the config loading code.

jdavcs committed:
Do not override schema defaults if they exist

Instead of _defaults_to('some val'), use schema default.

The following default values have been changed to use schema
defaults (I assume that was the intent):

key: old default / schema default
---------------------------------
genomespace_ui_url: none / https://gsui.genomespace.org/jsui/
communication_server_port: none / 7070
communication_server_host: none / http://localhost
persistent_communication_rooms: none / []
require_login: none / false
inactivity_box_content: none / schema text
registration_warning_message: none  / schema text
welcome_url: none / /static/welcome.html

I did not modify this if the schema default was the opposite (true vs
false) or if it was overwritten in the config loading code.

jmchilton commented on lib/galaxy/managers/configuration.py:
```diff
@@ -29,67 +29,67 @@ def default_serializer(self, config, key):
-            'welcome_url'                       : _defaults_to(None),
-            'show_welcome_with_login'           : _defaults_to(True),
-            'cookie_domain'                     : _defaults_to(None),
+            'require_login'                     : _defaults_to(self.app.config.require_login),
```
This looks really confused right - since ``_defaults_to`` is what is used if ``config`` doesn't have the attribute but the default your filling in is an unprotected access of that exact value from ``config``. This would all read a lot better I think if ``_defaults_to(self.app.config.require_login)`` was just ``_required_attribute``

with 

```python
            def _required_attribute(i, k, **c):
                assert hasattr(i, k)
                return getattr(i, k)
```

jdavcs replied:
@jmchilton  I don't think access to such attributes should be protected: it it's defined in the schema, the config object is guaranteed to have that attribute. Consider these 2 lines: 
1) https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L226 - this is where the value for the future attribute is loaded: it's either the schema default, or `None` (for `require_login`, `data` is this dict: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/config_schema.yml#L1976 ), and ` None` is the default used in `get()`, of course.
2) https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L301 - that's where the attribute is created.

This was one of my goals with the config edits: if there is a `foo` in the schema, it should be always accessible as `config.foo`, regardless of whether a default is specified, or whether it has been overwritten with `None`, etc.

If it's *not* in the schema, then it is either (a) appropriate to override it with a default based on context; or (b) if it's the same across the codebase, it might belong in the schema.

Is this reasonable?

jmchilton replied:
I’m not suggesting you should use a protected access pattern here.

jmchilton replied:
```python
'require_login': _defaults_to(self.app.config.require_login),
```

causes _defaults_to to be applied to config right? So you're effectively saying:

```python
getattr(self.app.config, 'require_login', self.app.config.require_login)
```

better just to annotate that it has a required value in some fashion. My idea was:

```python
'require_login': _required_attribute,
```

jdavcs replied:
Ah, of course! Yes, this makes sense. I'll make the change. Thanks!

jdavcs committed:
Simplify ref to schema default

For consistency, this also falls back to schema default for `terms_url`,
`brand`, `remote_user_logout_href`, which may default to `None`.
Previously this defaulted to `''`, I don't think a default value of
`None` will cause any issues here.

jdavcs committed:
Handle special case of mailing_lists

The config property in the schema is `mailing_lists_url`. The serializer
key is `mailing_lists`.

I think this inconsistency should be resolved (normalize to schema key),
but not in this PR. (added this to my config-todo-list)

jdavcs commented:
unit test failure unrelated.

jmchilton commented:
Thanks for the cleanup.
